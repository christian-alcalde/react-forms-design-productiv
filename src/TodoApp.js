import React, { useState } from "react";
import { v4 as uuid } from "uuid";

import TopTodo from "./TopTodo";
import EditableTodoList from "./EditableTodoList";
import TodoForm from "./TodoForm";
import Todo from "./Todo";

/** App for managing a todo list.
 *
 * Props:
 * - initialTodos: possible array of [ todo, ... ]
 *
 * State:
 * - todos: array of [ todo, ... ]
 *
 * App -> TodoApp -> { TodoForm, EditableTodoList }
 */

function TodoApp() {
  const [todos, setTodos] = useState([]);

  /** add a new todo to list */
  function create(newTodo) {
    let todo = { ...newTodo, id: uuid() };
    setTodos((todos) => [...todos, todo]);
  }

  /** update a todo with updatedTodo */
  function update(updatedTodo) {
    setTodos((todos) =>
      todos.map((todo) =>
        todo.id === updatedTodo.id ? { ...updatedTodo } : todo
      )
    );
  }

  /** delete a todo by id */
  function remove(id) {
    setTodos((todos) => todos.filter((todo) => id !== todo.id));
  }

  return (
    <main className="TodoApp">
      <div className="row">
        <div className="col-md-6">
          <h3>Todos</h3>
          {!todos.length ? (
            <span className="text-muted">You have no todos.</span>
          ) : (
            <EditableTodoList todos={todos} update={update} remove={remove} />
          )}
        </div>
        <div className="col-md-6">
          <h3>Top Todo</h3>
          {!todos.length ? (
            <div className="mb-4">
            <span className="text-muted">No todos yet!</span>
            <br />
            </div>
          ) : (
            <section className="mb-4">
              <TopTodo todos={todos} />
            </section>
          )}
          <section>
            <h3 className="mb-3">Add NÃ¼</h3>
            <TodoForm handleSave={create} />
          </section>
        </div>
      </div>
    </main>
  );
}

export default TodoApp;
