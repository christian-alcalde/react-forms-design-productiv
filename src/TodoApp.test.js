import React from "react";
import { render, fireEvent, getByDisplayValue } from "@testing-library/react";
import TodoApp from "./TodoApp";

const testTodos = [
  {
    id: "1",
    title: "laundry",
    description: "testDescription",
    priority: 3,
  },
  {
    id: "2",
    title: "water plants",
    description: "testDescription2",
    priority: 1,
  },
  {
    id: "3",
    title: "fix bike",
    description: "testDescription3",
    priority: 2,
  },
];

it("renders without crashing", function () {
  render(<TodoApp initialTodos={testTodos} />);
});

it("matches snapshot", function () {
  const { container } = render(<TodoApp initialTodos={testTodos} />);
  expect(container).toMatchSnapshot();
});

it("shows message if no todos", function () {
  const { container } = render(<TodoApp initialTodos={[]} />);

  expect(container).toHaveTextContent("No todos yet!");
  expect(container).toHaveTextContent("You have no todos.");
});

it("create function adds new todo", function () {
  const { container, getByLabelText, screen, debug } = render(
    <TodoApp initialTodos={[]} />
  );

  expect(container).toHaveTextContent("No todos yet!");
  expect(container).toHaveTextContent("You have no todos.");

  const titleInput = getByLabelText("Title");
  const descriptionInput = getByLabelText("Description");
  const priorityInput = getByLabelText("Priority:");

  fireEvent.change(titleInput, { target: { value: "hello" } });
  fireEvent.change(descriptionInput, { target: { value: "hi" } });
  fireEvent.change(priorityInput, { target: { value: 1 } });

  const submitBtn = container.querySelector(".NewTodoForm-addBtn");
  fireEvent.click(submitBtn);
  debug(screen);
  // expect(screen.getByTestId("laundry")).toBeInTheDocument();
  expect(container.querySelectorAll("li").length).toEqual(1);

  // testing if inputs are empty
  expect(titleInput).toContainHTML("");
  expect(descriptionInput).toHaveTextContent("");
  expect(priorityInput).toHaveTextContent("Ultra-Über");

  expect(container).not.toHaveTextContent("No todos yet!");
  expect(container).not.toHaveTextContent("You have no todos.");
});

it("updates a todo", function () {
  const { container, getByDisplayValue } = render(
    <TodoApp initialTodos={testTodos} />
  );

  expect(container).toHaveTextContent("laundry");

  const editButton = container.querySelector(".EditableTodo-toggle");
  fireEvent.click(editButton);

  const titleInput = getByDisplayValue("laundry");
  const descriptionInput = getByDisplayValue("testDescription");
  const priorityInput = getByDisplayValue("Ultra-Über");

  fireEvent.change(titleInput, { target: { value: "hello" } });
  fireEvent.change(descriptionInput, { target: { value: "hi" } });
  fireEvent.change(priorityInput, { target: { value: 1 } });

  const submitBtn = container.querySelector(".NewTodoForm-addBtn");
  fireEvent.click(submitBtn);

  expect(container).not.toHaveTextContent("laundry");
  expect(container).toHaveTextContent("hello");
  expect(container).toHaveTextContent("hi");
});
