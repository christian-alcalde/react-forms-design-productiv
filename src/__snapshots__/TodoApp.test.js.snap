// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<div>
  <main
    class="TodoApp"
  >
    <div
      class="row"
    >
      <div
        class="col-md-6"
      >
        <h3>
          Todos
        </h3>
        <div>
          <ul>
            <li
              data-testid="laundry"
            >
              <div
                class="EditableTodo"
              >
                <div
                  class="mb-3"
                >
                  <div
                    class="float-end text-sm-end"
                  >
                    <button
                      class="EditableTodo-toggle btn-link btn btn-sm"
                    >
                      Edit
                    </button>
                    <button
                      class="EditableTodo-delBtn btn-link btn btn-sm text-danger"
                    >
                      Del
                    </button>
                  </div>
                  <div
                    class="Todo"
                  >
                    <div>
                      <b>
                        laundry
                      </b>
                       
                      <small>
                        (Priority:
                        3
                        )
                      </small>
                    </div>
                    <div>
                      <small>
                        testDescription
                      </small>
                    </div>
                  </div>
                </div>
              </div>
            </li>
            <li
              data-testid="water plants"
            >
              <div
                class="EditableTodo"
              >
                <div
                  class="mb-3"
                >
                  <div
                    class="float-end text-sm-end"
                  >
                    <button
                      class="EditableTodo-toggle btn-link btn btn-sm"
                    >
                      Edit
                    </button>
                    <button
                      class="EditableTodo-delBtn btn-link btn btn-sm text-danger"
                    >
                      Del
                    </button>
                  </div>
                  <div
                    class="Todo"
                  >
                    <div>
                      <b>
                        water plants
                      </b>
                       
                      <small>
                        (Priority:
                        1
                        )
                      </small>
                    </div>
                    <div>
                      <small>
                        testDescription2
                      </small>
                    </div>
                  </div>
                </div>
              </div>
            </li>
            <li
              data-testid="fix bike"
            >
              <div
                class="EditableTodo"
              >
                <div
                  class="mb-3"
                >
                  <div
                    class="float-end text-sm-end"
                  >
                    <button
                      class="EditableTodo-toggle btn-link btn btn-sm"
                    >
                      Edit
                    </button>
                    <button
                      class="EditableTodo-delBtn btn-link btn btn-sm text-danger"
                    >
                      Del
                    </button>
                  </div>
                  <div
                    class="Todo"
                  >
                    <div>
                      <b>
                        fix bike
                      </b>
                       
                      <small>
                        (Priority:
                        2
                        )
                      </small>
                    </div>
                    <div>
                      <small>
                        testDescription3
                      </small>
                    </div>
                  </div>
                </div>
              </div>
            </li>
          </ul>
        </div>
      </div>
      <div
        class="col-md-6"
      >
        <h3>
          Top Todo
        </h3>
        <section
          class="mb-4"
        >
          <div
            class="Todo"
          >
            <div>
              <b>
                water plants
              </b>
               
              <small>
                (Priority:
                1
                )
              </small>
            </div>
            <div>
              <small>
                testDescription2
              </small>
            </div>
          </div>
        </section>
        <section>
          <h3
            class="mb-3"
          >
            Add Nü
          </h3>
          <form
            class="NewTodoForm"
          >
            <div
              class="mb-3"
            >
              <input
                aria-label="Title"
                class="form-control"
                id="newTodo-title"
                name="title"
                placeholder="Title"
                value=""
              />
            </div>
            <div
              class="mb-3"
            >
              <textarea
                aria-label="Description"
                class="form-control"
                id="newTodo-description"
                name="description"
                placeholder="Description"
              />
            </div>
            <div
              class="mb-3 d-flex justify-content-between"
            >
              <div
                class="w-75 d-flex justify-content-between"
              >
                <label
                  class="d-inline-flex"
                  for="newTodo-priority"
                >
                  Priority:  
                </label>
                <select
                  class="form-control form-control-sm d-inline-flex"
                  id="newTodo-priority"
                  name="priority"
                >
                  <option
                    value="1"
                  >
                    Ultra-Über
                  </option>
                  <option
                    value="2"
                  >
                    Über
                  </option>
                  <option
                    value="3"
                  >
                    Meh
                  </option>
                </select>
              </div>
              <button
                class="btn-primary rig btn btn-sm NewTodoForm-addBtn"
              >
                Gø!
              </button>
            </div>
          </form>
        </section>
      </div>
    </div>
  </main>
</div>
`;
